File: update_quality.rb
require 'award'

def update_quality(awards)
  awards.each do |award|
    case award.name
    when 'Blue First' 
      update_blue_first_award!(award)
    when 'Blue Distinction Plus'
      update_blue_distinction_plus!(award)
    when 'Blue Compare'
      update_blue_compare!(award)
    when 'Blue Star'
      update_blue_star!(award)
    else 
      update_normal_award!(award)
    end
  end
end

def update_blue_distinction_plus!(award)
  # blue distinction plus awards are not modified
end

def update_blue_star!(award)
  award.expires_in -= 1

  if award.expires_in > 0    
    award.quality = [0, award.quality - 2].max
  else
    award.quality = [0, award.quality - 4].max
  end
end

def update_normal_award!(award)
  award.expires_in -= 1 

  if award.expires_in > 0
    award.quality = [0, award.quality - 1].max
  else
    award.quality = [0, award.quality - 2].max
  end
end

def update_blue_first_award!(award)
  award.expires_in -= 1 

  if award.expires_in > 0
    award.quality = [50, award.quality + 1].min
  else
    award.quality = [50, award.quality + 2].min
  end
end

def update_blue_compare!(award)
  award.expires_in -= 1 

  if award.expires_in < 0
    award.quality = 0
  elsif award.expires_in < 5
    award.quality = [50, award.quality + 3].min
  elsif award.expires_in < 10
    award.quality = [50, award.quality + 2].min
  else
    award.quality = [50, award.quality + 1].min
  end
end

File: update_quality.py
def update_quality(awards):
    for award in awards:
        if award.name == 'Blue First':
            update_blue_first_award(award)
        elif award.name == 'Blue Distinction Plus':
            update_blue_distinction_plus_award(award)
        elif award.name == 'Blue Compare':
            update_blue_compare_award(award)
        elif award.name == 'Blue Star':
            update_blue_star_award(award)
        else:
            update_normal_award(award)

def update_blue_star_award(award):
    award.expires_in -= 1

    if award.expires_in > 0    :
        award.quality = max(0, award.quality - 2)
    else:
        award.quality = max(0, award.quality - 4)

def update_blue_compare_award(award):
    award.expires_in -= 1 

    if award.expires_in < 0:
        award.quality = 0
    elif award.expires_in < 5:
        award.quality = min(50, award.quality + 3)
    elif award.expires_in < 10:
        award.quality = min(50, award.quality + 2)
    else:
        award.quality = min(50, award.quality + 1)

def update_blue_distinction_plus_award(award):
    # blue distinction plus awards are not modified
    pass

def update_blue_first_award(award):
    award.expires_in -= 1

    if award.expires_in > 0:
        award.quality = min(50, award.quality + 1)
    else:
        award.quality = min(50, award.quality + 2)

def update_normal_award(award):
    award.expires_in -= 1

    if award.expires_in > 0:
        award.quality = max(0, award.quality - 1)
    else:
        award.quality = max(0, award.quality - 2)


